internal/telemetry/client.go

```diff
@@ -0,0 +1,407 @@
+	metricKindCount metricKind = "count"
+)
+
+type metric struct {
+	name  string
+	kind  metricKind
+	value float64
+	// Unix timestamp
+	ts     float64
+	tags   []string
+	common bool
+}
+
+// TODO: Can there be identically named/tagged metrics with a "common" and "not
+// common" variant?
+
+func newmetric(name string, kind metricKind, tags []string, common bool) *metric {
+	return &metric{
+		name:   name,
+		kind:   kind,
+		tags:   append([]string(nil), tags...),
+		common: common,
+	}
+}
+
+func metricKey(name string, tags []string) string {
+	return name + strings.Join(tags, "-")
+}
+
+// Gauge sets the value for a gauge with the given name and tags. If the metric
+// is not language-specific, common should be set to true
+func (c *Client) Gauge(name string, value float64, tags []string, common bool) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if !c.started {
+		return
+	}
+	key := metricKey(name, tags)
+	m, ok := c.metrics[key]
+	if !ok {
+		m = newmetric(name, metricKindGauge, tags, common)
+		c.metrics[key] = m
+	}
+	m.value = value
+	m.ts = float64(time.Now().Unix())
+	c.newMetrics = true
+}
+
+// Count adds the value to a count with the given name and tags. If the metric
+// is not language-specific, common should be set to true
+func (c *Client) Count(name string, value float64, tags []string, common bool) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if !c.started {
+		return
+	}
+	key := metricKey(name, tags)
+	m, ok := c.metrics[key]
+	if !ok {
+		m = newmetric(name, metricKindCount, tags, common)
+		c.metrics[key] = m
+	}
+	m.value += value
+	m.ts = float64(time.Now().Unix())
+	c.newMetrics = true
+}
+
+// Flush sends any outstanding telemetry messages and aggregated metrics to be
+// sent to the backend. Requests are sent in the background
+func (c *Client) Flush() {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	c.flush()
+}
+
+// actual flush implementation, should be called with c.mu locked
+func (c *Client) flush() {
+	submissions := make([]*Request, 0, len(c.requests)+1)
+	if c.newMetrics {
+		c.newMetrics = false
+		r := c.newRequest(RequestTypeGenerateMetrics)
+		payload := &Metrics{
+			Namespace:   c.Namespace,
+			LibLanguage: "golang",
+			LibVersion:  version.Tag,
+		}
+		for _, m := range c.metrics {
+			s := Series{
+				Name:   m.name,
+				Type:   string(m.kind),
+				Tags:   m.tags,
+				Common: m.common,
+			}
+			s.Points = [][2]float64{{m.ts, m.value}}
+			payload.Series = append(payload.Series, s)
+		}
+		r.Payload = payload
+		submissions = append(submissions, r)
+	}
+
+	// copy over requests so we can do the actual submission without holding
+	// the lock. Zero out the old stuff so we don't leak references
+	for i, r := range c.requests {
+		submissions = append(submissions, r)
+		c.requests[i] = nil
+	}
+	c.requests = c.requests[:0]
+
+	go func() {
+		for _, r := range submissions {
+			err := c.submit(r)
+			if err != nil {
+				c.log("telemetry submission failed: %s", err)
+				continue
```

Comment:
NIT: This `continue` is a no-op, right? Was it put here to communicate intend? If yes, I'd probably prefer it as a `// continue` comment.

Reply:
Yep, I had it there as a placeholder and didn't remove it, but I'll remove it now